File: sdks/php/src/HookRelayException.php
<?php

declare(strict_types=1);

namespace HookRelay;

class HookRelayException extends \RuntimeException
{
    private string $errorCode;

    public function __construct(string $message, int $statusCode = 0, string $errorCode = 'UNKNOWN')
    {
        parent::__construct($message, $statusCode);
        $this->errorCode = $errorCode;
    }

    public function getErrorCode(): string
    {
        return $this->errorCode;
    }
}

class AuthenticationException extends HookRelayException
{
    public function __construct(string $message = 'Unauthorized: invalid or missing API key')
    {
        parent::__construct($message, 401, 'UNAUTHORIZED');
    }
}

class NotFoundException extends HookRelayException
{
    public function __construct(string $message = 'Resource not found')
    {
        parent::__construct($message, 404, 'NOT_FOUND');
    }
}

class RateLimitException extends HookRelayException
{
    public function __construct(string $message = 'Rate limit exceeded')
    {
        parent::__construct($message, 429, 'RATE_LIMIT_EXCEEDED');
    }
}

class ValidationException extends HookRelayException
{
    public function __construct(string $message = 'Bad request')
    {
        parent::__construct($message, 400, 'BAD_REQUEST');
    }
}

class PayloadTooLargeException extends HookRelayException
{
    public function __construct(string $message = 'Payload too large')
    {
        parent::__construct($message, 413, 'PAYLOAD_TOO_LARGE');
    }
}

File: sdks/java/src/main/java/com/hookrelay/HookRelayException.java
package com.hookrelay;

/**
 * Base exception for HookRelay API errors.
 */
public class HookRelayException extends RuntimeException {
    private final int statusCode;
    private final String errorCode;

    public HookRelayException(String message, int statusCode, String errorCode) {
        super(message);
        this.statusCode = statusCode;
        this.errorCode = errorCode;
    }

    public int getStatusCode() {
        return statusCode;
    }

    public String getErrorCode() {
        return errorCode;
    }

    public static class AuthenticationException extends HookRelayException {
        public AuthenticationException(String message) {
            super(message, 401, "UNAUTHORIZED");
        }
    }

    public static class NotFoundException extends HookRelayException {
        public NotFoundException(String message) {
            super(message, 404, "NOT_FOUND");
        }
    }

    public static class RateLimitException extends HookRelayException {
        public RateLimitException(String message) {
            super(message, 429, "RATE_LIMIT_EXCEEDED");
        }
    }

    public static class ValidationException extends HookRelayException {
        public ValidationException(String message) {
            super(message, 400, "BAD_REQUEST");
        }
    }

    public static class PayloadTooLargeException extends HookRelayException {
        public PayloadTooLargeException(String message) {
            super(message, 413, "PAYLOAD_TOO_LARGE");
        }
    }
}
